proposals/0613-ghc2024.rst

```diff
@@ -33,6 +33,47 @@ that ``GHC2021`` comprises (`official list <https://downloads.haskell.org/ghc/la
 * `LambdaCase`_
 * `RoleAnnotations`_
 
+Introduction of ``GHC2024``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The `GHC20xx process proposal
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0372-ghc-extensions.rst#proposed-change-specification>`_
+specifies that
+
+    When ``ghc`` is used without an explicit language choice, the latest
+    ``GHC20xx`` known to GHC is used. This applies in particular to uses of
+    ``ghci``.
+
+However, switching from ``GHC2021`` to ``GHC2024`` may break existing programs,
+primarily because of ``MonoLocalBinds``.  While Cabal packages will usually
+specify a ``default-language`` and hence be unaffected, this is not guaranteed.
+Moreover, some users compile programs using ``ghc`` directly, so a change in the
+default language edition used by ``ghc`` may cause backwards incompatibility for
+those users.
+
+Thus ``GHC2024`` will be introduced in a phased manner:
+
+* GHC 9.10 will add support for ``GHC2024`` but the default language edition
+  will remain ``GHC2021`` (in both ``ghc`` and ``ghci``).
+
+* In addition, GHC 9.10 and later releases will emit a warning (controlled by
+  ``-Wunspecified-language`` and in ``-Wdefault``) when compiling a ``.hs``
+  file without an explicit choice of language edition.  (Compiling a ``.hs``
+  file includes loading it into ``ghci``, but using ``ghci`` interactively
+  without specifying a language edition will not emit a warning.)
+
+* The next major release of GHC after 9.10 will switch the default language
+  edition to ``GHC2024``.
+
+This will give users warning of the impending change, and encourage them to
+specify a language edition explicitly, so that future changes of default
+language edition do not affect their code.  It does mean that throw-away uses of
+``ghc M.hs`` will emit a warning by default; one way to avoid this is to define
+an alias, e.g. ``alias ghc21='ghc -XGHC2021'``.  Users of ``ghci`` may be
```

Comment:
I'm conflicted.

I don't like this. But script calling ghc are likely to be the most common reason for no language edition being specified. So we need this behaviour for the proposal to be useful.

Is it worth the developer time that we'd put in it? I guess I'm kind of against this warning then.

Reply:
I'm against changing the default without informing users it will change.

If people don't read the warnings, it will break either way, not or later.
For those that _do_ read the warnings (and this might just come from running runghc or building something by hand, and those looking at (new) warnings in CI), this will help them make decisions about upgrading and adjusting easier.

The work involved for downstream users is the same, but the abrupt breakage is much more hostile than a gentle warning informing about the upcoming breakage.

Reply:
> I'm against changing the default without informing users it will change.

Is informing users in the user's guide / release notes / release announcement adequate? What if we ship 9.10 with `GHC2024` available but not the default, and change the default in the subsequent release, accompanying both releases with publicity for this plan (but not warnings by default)?

I realise that not everyone reads the release notes. But not everyone reads warnings either! And the alternative means warning for every one-off module forever, for the corner cases where somebody uses `ghc` (but not `cabal`) for a long-lived script and doesn't specify the language edition.